### PR TITLE
Add `bundle fund` command 

### PR DESCRIPTION
**Pull request by gjtorikian**
Migrated from https://github.com/rubygems/bundler/pull/7668

* * *

## What was the end-user or developer problem that led to this PR?

This PR implements the `bundle fund` proposal in [this accepted RFC](https://github.com/rubygems/rfcs/blob/aa031d3774a4ce6cb06d317ab03ea5cd161da498/text/0000-add-funding-uri.md).

## What is your fix for the problem, implemented in this PR?

The rationale and technical details for the `bundle fund` command are explained in the RFC, but roughly:

* A new metadata key, `funding_uri`, can be specified, that points to a funding page for a gem
* On a successful `bundle install` or `bundle update`, a message will be printed out that indicates gems which require support: `4 gems you depend on are looking for funding`
* `bundle fund` will list out all those gems, with their funding links
* `bundle info` will include the `funding_uri` in its printout
* You can restrict `bundle fund` to only list information for a specific group
* Dependencies of dependencies are currently not listed in `bundle fund`; this is [by design](https://github.com/rubygems/rfcs/blob/aa031d3774a4ce6cb06d317ab03ea5cd161da498/text/0000-add-funding-uri.md#unresolved-questions)

### CODE REVIEW

**Comment by gjtorikian** [comment]
Just bumping this to hope to get some attention to it. ✌️

**Comment by simi** [comment]
Overall looks good to me. I have two suggestions:

1. Once merged, promote `funding_uri` at https://guides.rubygems.org/specification-reference/#metadata as well.
2. What about to make it part of `bundle gem` gemspec template?

**Comment by gjtorikian** [comment]
>     1. Once merged, promote `funding_uri` at https://guides.rubygems.org/specification-reference/#metadata as well.

Sure, and individual `rubygems.org/gems` pages too.

>     2\. What about to make it part of `bundle gem` gemspec template?

From [the RFC](https://github.com/rubygems/rfcs/pull/23), I didn't want to do this. Maybe in a future iteration?

**Comment by gjtorikian** [comment]
Hey, just bumping this...again.

**Comment by deivid-rodriguez** [comment]
@gjtorikian I'm so sorry nobody stepped in to review this before. Thanks so much for being so patient.

I'm going to put this on my list for this week. If I haven't reviewed this by next Monday, please ping me again.

**Comment by deivid-rodriguez** [comment]
@gjtorikian Could you rebase this PR, it's 1289 commits behind current HEAD, I'd like to make sure it's green before reviewing. I can also do that for you as well if you're busy.

**Comment by gjtorikian** [comment]
@deivid-rodriguez Ok, I've rebased.

**Comment by deivid-rodriguez** [comment]
Tests are failing. It's because we removed the `bundle!` test helper. You should use `bundle` now.

**Comment by gjtorikian** [comment]
Ah, ok. Done!

**Comment by deivid-rodriguez** [comment]
Hello @gjtorikian!

I just read the RFC, it looks really really good :heart:. I agree with everything, including the potentially most controversial parts like not adding an `opt-out` flag to skip the `<n> gems you depend on are looking for funding` message. It's just a small unintrusive line, I don't think people will be bothered by it.

Regarding the message, what do you think about tweaking it to "<n> gems you **directly** depend on are looking for funding". That way we give some clue that they might be the only code you depend on that it's looking for funding.

Regarding the functionality, it looks all good other than that little tweak. I'll go ahead with reviewing the implementation now.

**Comment by gjtorikian** [comment]
@deivid-rodriguez That makes complete sense to me, and I've tweaked the message. I also removed the `!` as it felt a bit too enthusiastic.

File: bundler/spec/update/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/update/gems/fund_spec.rb:13]: This `config` seems unused? Maybe initially you meant to make this configurable?

File: bundler/spec/update/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/update/gems/fund_spec.rb:22]: It seems like it's better to move this shared spec inline since there's only one spec using it?

Comment (deivid-rodriguez) [reply bundler/spec/update/gems/fund_spec.rb:22]: So, this will repeat the same message every time `bundle update` is run, just like if `bundle install` had been run, right? I was thinking if this should only inform about funding of gems that were explicitly updated. But it sounds tricky and potentially misleading. This seems fine.

File: bundler/spec/install/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/fund_spec.rb:57]:
```suggestion
      it "should display the fund message after installing" do
```

File: bundler/spec/install/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/fund_spec.rb:72]:
```suggestion
      it "should display the fund message if repo is updated" do
```

File: bundler/spec/install/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/fund_spec.rb:98]:
```suggestion
      it "should still display the fund message if repo is not updated" do
```

**Review by deivid-rodriguez — COMMENTED** [review]
Made a first pass. Looking good!

**Review by gjtorikian — COMMENTED** [review]
Thanks @deivid-rodriguez ! Everything made sense and I've made the changes.

**Comment by gjtorikian** [comment]
So many test issues 🙈! But all good now.

File: bundler/lib/bundler/cli/common.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli/common.rb:17]: Actually what I meant is that this parameter is not really needed, so I think we can remove it plus all the changes in `bundler/lib/bundler/installer.rb`, and `bundler/lib/bundler/paralell_installer.rb`.

File: bundler/lib/bundler/cli/common.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli/common.rb:22]: Can this `nil?` actually happen? Also, even if it could happen, I guess you wouldn't want to early `return 0`, but instead make the result of the block invokation to be `false`? If we can't find a test case where this is needed, I'd rather to make the `nil?` check at all.

File: bundler/lib/bundler/cli/fund.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli/fund.rb:23]: I copied my `count` block above from here, and you're not checking for `spec.nil?` here. I guess it can't be `nil`.

File: bundler/lib/bundler/cli/fund.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli/fund.rb:17]: I think this is equivalent to `Bundler.definition.dependencies_for(groups)`.

File: bundler/spec/commands/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/commands/fund_spec.rb:43]: Super picky comment, but I prefer to use a "present tense" style like the other specs in this file rather than a "should style". So I'd go with

```suggestion
    it "prints fund message for only specified group gems" do
```

File: bundler/spec/install/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/fund_spec.rb:6]: Same thing about should vs present style for this spec and the others in this file.

File: bundler/spec/install/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/fund_spec.rb:14]: Unless you want to be explicit for a readabilty reason (maybe in the last spec in this file it makes sense to keep them separate), use `install_gemfile` over `gemfile` +  `bundle :install`.

File: bundler/spec/update/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/update/gems/fund_spec.rb:4]: This `let` can go too.

**Review by deivid-rodriguez — COMMENTED** [review]
I made another pass, mainly little style issues now.

File: bundler/spec/update/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/update/gems/fund_spec.rb:27]: Let's say

```suggestion
    it "displays fund message" do
```

since the current wording makes it unclear whether information is displayed for all gems or only the ones explicitly updated.

File: bundler/spec/update/gems/fund_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/update/gems/fund_spec.rb:16]: I see this is also coming from `spec/update/gems/post_install_spec.rb` but it doesn't make sense since `bundle update` is being run for all gems, not with a specific gem argument. So we could say

```suggestion
  context "when gems are updated" do
```

or something like that.

File: bundler/lib/bundler/cli/common.rb
Comment (gjtorikian) [reply bundler/lib/bundler/cli/common.rb]: Whoops!

File: bundler/lib/bundler/cli/common.rb
Comment (gjtorikian) [reply bundler/lib/bundler/cli/common.rb]: It can `nil`, but only in the test suites—I had tests failing without this check. I no longer have access to those runs, but perhaps they've been kept in the GitHub Actions logs? I guess what I could/should have done was to just fix the test suite. What do you think?

**Comment by gjtorikian** [comment]
@deivid-rodriguez Ok, I have figured out what was going on with that `nil` check. Here is the test which was causing the issue:

https://github.com/rubygems/rubygems/blob/5d2b50b02b73938611754602798522a8eba71447/bundler/spec/commands/exec_spec.rb#L271-L286

Well, it turns out, `current_dependencies` will _not_ respect that `--without`, while `specs` does: https://github.com/gjtorikian/rubygems/blob/2517d795f61057b64f7daa391e94b91c2fe6ca4f/bundler/lib/bundler/cli/common.rb#L18-L21

So the `current_dependencies` array we iterate over has more values than the `current_specs` list, because of the groups not excluded.

To fix this, I looked around the code, and honestly the easiest solution was to move `requested_dependencies` out of `private` and into `public`. `requested_specs` was already public, so I see no reason for `requested_dependencies` to have been kept public (other than perhaps its accidental inclusion under `private`--thanks Ruby!). [Here is the blame where `requested_dependencies` was made `private`](https://github.com/gjtorikian/rubygems/blame/1cdc4f03818275840eb4d575e25c95642a91dc1f/bundler/lib/bundler/definition.rb#L907) and [the commit](8307823d80925a1393a19dcf9c4eebf7f9554208) made over 10 years ago doesn't appear to have any reason for it to be private, let alone a strong one.

If this is unacceptable (changes to the API for example) I could do some sort of `groups  = Array(options[:group]).map(&:to_sym)` manipulation but it would probably be messier and not DRY.

**Comment by gjtorikian** [comment]
Ah yeah, seems that the `older-rubygems-bundler / older_rubygems_bundler (2.3.8, v2.6.14)` suite didn't like that. I'll pause here and wait for some guidance on how to handle it, then. Too many force pushes already. 🤪

File: bundler/lib/bundler/cli/common.rb
Comment (deivid-rodriguez) [reply bundler/lib/bundler/cli/common.rb]: Mmm, not sure, maybe undo the change and let's see what fails now.

**Comment by deivid-rodriguez** [comment]
Ok, I'll pick this up tomorrow again and check this!

File: bundler/lib/bundler/installer.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/installer.rb:18]: All the changes in this file can go now, right?

File: bundler/lib/bundler/installer/parallel_installer.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/installer/parallel_installer.rb:9]: Same with all the changes in this file.

File: bundler/lib/bundler/cli/fund.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli/fund.rb:17]: Tests failed because I suggested this method which is also private, I guess it's fine to make it public too.

**Review by deivid-rodriguez — COMMENTED** [review]
This feels like almost ready now!

File: bundler/lib/bundler/installer.rb
Comment (gjtorikian) [reply bundler/lib/bundler/installer.rb:18]: Whoops, yes, that was lazy of me. 🙈

File: bundler/lib/bundler/cli/fund.rb
Comment (gjtorikian) [reply bundler/lib/bundler/cli/fund.rb:17]: I think so too; `specs_for` is already also public.

**Comment by deivid-rodriguez** [comment]
That `nil` is still causing issues. I'll clone the repo later and figure it out.

**Comment by deivid-rodriguez** [comment]
Hello again @gjtorikian! I had a look at this. The problem is that `Bundler.definition.requested_specs` returns "not materialized" specs. That essentially means specs not relevant for the current platform are not filtered out. Your initial implementation indirectly used `Bundler.definition.specs` which materializes all the specs relevant to the requested groups:

https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/definition.rb#L180

So I think changing `Bundler.definition.requested_specs` to `Bundler.definition.specs` should get things back to working.

**Comment by gjtorikian** [comment]
@deivid-rodriguez Looks like that did it!

File: bundler/lib/bundler/cli/common.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli/common.rb:18]: Another nit, but for simmetry, could we avoid the `map` here, and instead do `dep.name` inside the `count` block?

Comment (deivid-rodriguez) [reply bundler/lib/bundler/cli/common.rb:18]: Also, I'm wondering whether we should actually use `requested_dependencies` here, since it kind of makes the message inaccurate because of taking groups into account. So, if your current environment is using, say, `without production`, it will omit to include the `production` gems in the count. But technically you still depend on those gems (they're pretty critical in this case, actually :sweat_smile:).

**Review by deivid-rodriguez — COMMENTED** [review]
Another two minor comments.

File: bundler/lib/bundler/cli/common.rb
Comment (gjtorikian) [reply bundler/lib/bundler/cli/common.rb]: > Also, I'm wondering whether we should actually use `requested_dependencies` here, since it kind of makes the message inaccurate because of taking groups into account.

I think I agree with you, and made that change!

**Comment by gjtorikian** [comment]
I have no idea why Windows builds are now broken. 😞

**Comment by deivid-rodriguez** [comment]
It sounds like a similar problem as before of not filtering out specs not relevant to the platform. If instead of 928591a5fe31bba05122d660f82cea6235679b8c, you change `current_dependencies = Bundler.definition.requested_dependecies` with `current_dependencies = Bundler.definition.current_dependencies`, does it work?

Also, could you rebase again since you are at it?

**Comment by gjtorikian** [comment]
> does it work?

It does not. I can reproduce the same `undefined method `metadata' for nil:NilClass` locally (on macOS). Perhaps `specs` does not take platforms into consideration? 

> Also, could you rebase again since you are at it?

Yes, I'll do this once we're all green.

**Comment by deivid-rodriguez** [comment]
mmm. Interesting. I cloned the PR, made the modification and run the relevant specs on Linux and they passed. Theorically `definition.specs` should filter out specs not relevant for the current platform, and avoid `nil` results. Can you still push the modification so that I can see the failures?

**Comment by deivid-rodriguez** [comment]
What if using the same reasoning as with groups before, we don't do any platform filtering either? Maybe

```
current_dependencies = Bundler.definition.dependencies
current_specs = Bundler.definition.resolve
```

As you can see, this is all a bit of a mess :sweat_smile:.

**Comment by gjtorikian** [comment]
> Can you still push the modification so that I can see the failures?

Ok, I pushed it!

**Comment by deivid-rodriguez** [comment]
Oh wow, is there anything my suggestion doesn't break? :laughing:

Ok, let me check this.

**Comment by deivid-rodriguez** [comment]
Ok @gjtorikian, sorry for the back and forth here.

The problem is that `specs` already does the filtering of excluding not configured groups, so we can't get the funding information for those from there. We could try `resolve.materialize(current_dependencies)` instead to get _all_ materialized specs (without any group filtering), but in order for that to work, it turns out we'd need to actually install those specs, and that didn't happen because groups were filtered out in the first place.

So, since I don't see an easy fix for this, I propose we go back to https://github.com/rubygems/rubygems/pull/3390#discussion_r495939587 and address that instead by further tweaking the message to

> \<n\> _installed_ gems you directly depend on are looking for funding.

to make it more clear that any gem exclusions happening before gem installation, for example `without production`, are not included in the count.

**Comment by gjtorikian** [comment]
🎆

**Review by deivid-rodriguez — APPROVED** [review]
Hi!

I'm aware that at this point I'm probably just being annoying, but technically the commit that needs `requested_dependencies` to be made public is the second, so the change making that method public belongs in there :sweat_smile:.

Other than that, I'd like to thank you so much for your infinite patience and for your polite well-spaced pings along these past months :purple_heart:.

**Comment by gjtorikian** [comment]
Heh, no worries. It's fixed up!

I am just excited to be able to add this and get some visibility for support. I have another change pending for visualizing this metadata on rubygems.org next. 😄

**Comment by deivid-rodriguez** [comment]
Hei, that's awesome to hear. This will make a better ecosystem overall.

Regarding my last request, I don't think it's actually fixed because `bundle fund` requires `dependencies_for` to be made public, so that part should go with the first commit, while the post_install method requires `requested_dependencies` to be made public, so that part should be included in the second commit. The reason I ask for this is not only for a better git history, but also if I ever need to bisect something, I'd like each individual commit to pass specs individually.

**Comment by gjtorikian** [comment]
> The reason I ask for this is not only for a better git history, but also if I ever need to bisect something, I'd like each individual commit to pass specs individually

It's all good! It's your project, I'm just committing to it. 😆 

Re-rebased and pushed.

**Comment by deivid-rodriguez** [comment]
Awesome, thanks so much @gjtorikian!! :tada:

**Comment by simi** [comment]
:partying_face:

**Comment by gjtorikian** [comment]
As a maintainer I know how problematic the next question is but I am compelled to ask: when can we expect a new release with these changes? 🙈

**Comment by deivid-rodriguez** [comment]
We hope to release bundler 2.2.0.rc.2 next week including this feature.

**Pull request merged**
